### PR TITLE
ctfeexpr: support reinterpreting cast when switching safe/pure/nothrow in both directions

### PR DESCRIPTION
**Pull request by ljmf00**
Signed-off-by: Luís Ferreira <contact@lsferreira.net>

### CODE REVIEW

**Comment by UplinkCoder** [comment]
allowing this cast breaks the type system

**Comment by UplinkCoder** [comment]
Where's the bugreport which requests this?

**Comment by ljmf00** [comment]
> allowing this cast breaks the type system

Yes. Although, AFAIK CTFE engine can handle these cases. Some functions on standard library use `assumePure` that does this type of casts. I can do further testing to see if this is viable.

**Comment by ljmf00** [comment]
> Where's the bugreport which requests this?

There's no bug report for this, atm.

**Comment by MoonlightSentinel** [comment]
> There's no bug report for this, atm.

Please create a new one or add a changelog entry

File: test/compilable/reinterpretctfe.d [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// REQUIRED_ARGS:
+int nonPureFunc()
+{
+    return 2;
+}
+
+void main()
+{
+    static immutable ctfeI = (cast(int function() pure)&nonPureFunc);
```
Comment (MoonlightSentinel) [inline test/compilable/reinterpretctfe.d:9]: Mabye extend this test case to actually call the casted function pointer? (just to be sure)

**Comment by MoonlightSentinel** [comment]
Added a bugzilla reference and slightly extended the test case

**Review by thewilsonator — APPROVED** [review]

**Comment by ljmf00** [comment]
> Added a bugzilla reference and slightly extended the test case

I'm so sorry for not answering quickly. As I mentioned previously here, https://github.com/dlang/dlang.org/pull/2946#issuecomment-851440669, I'm very busy due to school stuff.

**Comment by MoonlightSentinel** [comment]
No worries, I just wanted to have these changes to enable dlang/druntime#3494.

**Comment by tom-tan** [comment]
Can we extend the same idea to other interpreting casts?
For example, CTFEability of casting from immutable to mutable is one of blockers of [issue 21405](https://issues.dlang.org/show_bug.cgi?id=21405).

**Comment by ljmf00** [comment]
> Can we extend the same idea to other interpreting casts?
> For example, CTFEability of casting from immutable to mutable is one of blockers of [issue 21405](https://issues.dlang.org/show_bug.cgi?id=21405).

I can look at this when I have some time.

**Pull request merged**